core/src/com/unciv/logic/automation/unit/EspionageAutomation.kt

```diff
@@ -1,33 +1,83 @@
 package com.unciv.logic.automation.unit
 
+import com.unciv.logic.city.City
 import com.unciv.logic.civilization.Civilization
+import com.unciv.models.Spy
 import com.unciv.models.SpyAction
+import kotlin.random.Random
 
-object EspionageAutomation {
+class EspionageAutomation(val civInfo: Civilization) {
+    private val civsToStealFrom: List<Civilization> by lazy {
+        civInfo.getKnownCivs().filter {otherCiv -> otherCiv.isMajorCiv() && otherCiv.cities.any { it.getCenterTile().isVisible(civInfo) }
+            && civInfo.espionageManager.getTechsToSteal(otherCiv).isNotEmpty() }.toList()
+    }
 
-    fun automateSpies(civInfo: Civilization) {
-        val civsToStealFrom: List<Civilization> by lazy { 
-            civInfo.getKnownCivs().filter {otherCiv -> otherCiv.isMajorCiv() && otherCiv.cities.any { it.getCenterTile().isVisible(civInfo) } 
-                && civInfo.espionageManager.getTechsToSteal(otherCiv).isNotEmpty() }.toList()
-        }
+    private val getCivsToStealFromSorted: List<Civilization> =
+        civsToStealFrom.sortedBy { otherCiv -> civInfo.espionageManager.spyList
+            .count { it.isDoingWork() && it.getLocation()?.civ == otherCiv }
+        }.toList()
+
+    private val cityStatesToRig: List<Civilization> by lazy {
+        civInfo.getKnownCivs().filter { otherCiv -> otherCiv.isMinorCiv() && otherCiv.knows(civInfo) && !civInfo.isAtWarWith(otherCiv) }.toList()
+    }
 
-        val getCivsToStealFromSorted: List<Civilization> =
-            civsToStealFrom.sortedBy { otherCiv -> civInfo.espionageManager.spyList
-                .count { it.isDoingWork() && it.getLocation()?.civ == otherCiv } 
-            }.toList()
-
-        for (spy in civInfo.espionageManager.spyList) {
-            if (spy.isDoingWork()) continue
-            if (civsToStealFrom.isNotEmpty()) {
-                // We want to move the spy to the city with the highest science generation
-                // Players can't usually figure this out so lets do highest population instead
-                spy.moveTo(getCivsToStealFromSorted.first().cities.filter { it.getCenterTile().isVisible(civInfo) }.maxByOrNull { it.population.population })
+    fun automateSpies() {
+        val spies = civInfo.espionageManager.spyList
+        val spiesDoingCounterIntelligence = spies.count { it.action == SpyAction.CounterIntelligence }
+        val spiesToMove = spies.filter { it.isAlive() && 
+            (!it.isDoingWork() || (it.action == SpyAction.CounterIntelligence && spiesDoingCounterIntelligence > 2))}
```

Comment:
This is a complex boolean operation, we can make it slightly simpler by replacing the && with .filter{...},filter{...}

Reply:
Actually, I can remove the second counter-intelligence clause. I had it in there before I set a time limit for how long counter-intelligence counts as work for.